# Lab 4

1. 提交作業之前，建議可以先點選上方工具列的**Kernel**，再選擇**Restart & Run All**，檢查一下是否程式跑起來都沒有問題，最後記得儲存。
2. 請先填上下方的姓名(name)及學號(stduent_id)再開始作答，例如：
```python
name = "我的名字"
student_id= "B06201000"
```
3. 演算法的實作可以參考[lab-4](https://yuanyuyuan.github.io/itcm/lab-4.html), 有任何問題歡迎找助教詢問。
4. **Deadline: 11/20(Wed.)**

In [1]:
name = "黃宇文"
student_id = "B06201029"

---

# Exercise 2
---
### Let $I(f)$ be a define integral defined by
### $$I(f) = \int_0^1 f(x) dx,$$
### and consider the quadrature formula
### $$\hat{I}(f) = \alpha_1 f(0) + \alpha_2 f(1) + \alpha_3 f'(0) \quad \quad \quad (*)$$
### for approximation of $I(f)$.

---
### Part 1. 

#### Determine the coefficients $\alpha_j$ for $j = 1,\,2,\,3$ in such a way that $\hat{I}$ has the degree of exactness $r = 2$. Here the degree of exactness $r$ is to find $r$ such that
#### $$\hat{I}(x^k)=\hat{I}(x^j) \quad \text{for} \quad k = 0,\,1,\dots,\,r \quad \text{and} \quad \hat{I}(x^j) \neq I(x^j) \quad \text{for} \quad j > r, $$
#### where $x^j$ denote the $j$-th power of $x$.
---

Derive the values of $\alpha_1, \alpha_2, \alpha_3$ in $(*)$.  You need to write down the detail in the cell below with Markdown/LaTeX. 

For $f(x)=1$,
\begin{align}
\hat{I}(f) = \int_0^1 \! 1 \, \mathrm{d}x &= \alpha_1(1) + \alpha_2(1) + \alpha_3(0) \\
1 &= \alpha_1 + \alpha_2
\end{align}
\
For $f(x)=x$,
\begin{align}
\hat{I}(f) = \int_0^1 \! x \, \mathrm{d}x &= \alpha_1(0) + \alpha_2(1) + \alpha_3(1) \\
\frac{1}{2} &= \alpha_2 + \alpha_3
\end{align}
\
For $f(x)=x^2$,
\begin{align}
\hat{I}(f) = \int_0^1 \! x^2 \, \mathrm{d}x &= \alpha_1(0) + \alpha_2(1) + \alpha_3(0) \\
\frac{1}{3} &= \alpha_2
\end{align}
\
\begin{align}
\alpha_1 &= 1 - \frac{1}{3} = \frac{2}{3} \\
\alpha_2 &= \frac{1}{3} \\
\alpha_3 &= \frac{1}{2} - \frac{1}{3} = \frac{1}{6}
\end{align}

Fill in the tuple variable `alpha_1`,  `alpha_2`, `alpha_3` with your answer above. 

In [2]:
'''Hint: 
alpha_1 = ?
alpha_2 = ?
alpha_3 = ?
'''
# ===== 請實做程式 =====
alpha_1 = 2/3
alpha_2 = 1/3
alpha_3 = 1/6
# ====================

In [3]:
print("alpha_1 =", alpha_1)
print("alpha_2 =", alpha_2)
print("alpha_3 =", alpha_3)

alpha_1 = 0.6666666666666666
alpha_2 = 0.3333333333333333
alpha_3 = 0.16666666666666666


---
### Part 2. 

#### Find an apppropriate expression for the error $E(f) = I(f) - \hat{I}(f)$, and write your process in the below cell with Markdown/LaTeX.

---

\begin{align}
\hat{I}(f) &= \frac{2}{3}f(0) + \frac{1}{3}f(1) + \frac{1}{6}f'(0) \\
I(f) &= \int_0^1 \! f(x) \, \mathrm{d}x \\
E(f) &= I(f) - \hat{I}(f) = \frac{1}{2!}\int_0^1 \! f^{(3)}(t)E\left((x-t)^2_+\right) \, \mathrm{d}t \\
\\
E\left((x-t)^2_+\right) &= \int_0^1 \! (x-t)^2_+ \, \mathrm{d}x - \hat{I}\left((x-t)^2_+\right) \\
&= \int_t^1 \! (x-t)^2 \, \mathrm{d}x - \hat{I}\left((x-t)^2_+\right) \\
&= \left[\frac{(x-t)^2}{3}\right]^1_{x=t} - \left(\frac{2}{3}(0-t)^2_+ + \frac{1}{3}(1-t)^2_+ + \frac{1}{3}(0-t)_+ \right) \\
&= \frac{(1-t)^3}{3} - \frac{(1-t)^2}{3} = -\frac{1}{3}t(1-t)^2\\
\\
\therefore E(f) &= -\frac{1}{6}\int_0^1 \! t(1-t)^2f^{(3)}(t) \, \mathrm{d}t 
\end{align}

---
### Part 3. 
### Compute
###  $$\int_0^1 e^{-\frac{x^2}{2}} dx$$
###  using quadrature formulas $(*)$, the Simpson's rule and the Gauss-Legendre formula in the case $n = 1$. Compare the obtained results.

### Part 3.1

Import necessary libraries

In [4]:
import numpy as np
import matplotlib.pyplot as plt
from scipy.special.orthogonal import p_roots

### Part 3.2 

Define the function $f(x) = e^{-\frac{x^2}{2}}$ and its derivative.

In [5]:
def f(x):
    # ===== 請實做程式 =====
    return np.exp(-x**2/2)
    # ====================
    
def d_f(x):
    # ===== 請實做程式 =====
    return np.exp(-x**2/2)*(-x)
    # ====================

Print and check your functions.

In [6]:
print('f(0) =', f(0))
print("f'(0) =", d_f(0))

f(0) = 1.0
f'(0) = 0.0


### Part 3.3

Compute 

$$\int_0^1 e^{-\frac{x^2}{2}} dx$$

with the formula $(*)$.

Fill your answer into the variable `approximation`.

In [7]:
# Hint: approximation = ...
# ===== 請實做程式 =====
approximation = alpha_1 * f(0) + alpha_2 * f(1) + alpha_3 * d_f(0)
# ====================

Run and check your answer. 

In [8]:
print("The result of the integral is", approximation)

The result of the integral is 0.8688435532375445


### Part 3.4 

Compute 

$$\int_0^1 e^{-\frac{x^2}{2}} dx$$

with Simpson's rule.

Implement Simpson's rule

In [9]:
def simpson(
    f,
    a,
    b,
    N=50
):
    '''
    Parameters
    ----------
    f : function
        Vectorized function of a single variable
    a , b : numbers
        Interval of integration [a,b]
    N : (even) integer
        Number of subintervals of [a,b]

    Returns
    -------
    S : float
        Approximation of the integral of f(x) from a to b using
        Simpson's rule with N subintervals of equal length.
    '''
    # ===== 請實做程式 =====
    if N % 2 == 1:
        raise ValueError('N must be an even integer.')
    
    dx = (b-a) / N
    x = np.linspace(a, b, N+1)
    y = f(x)
    S = dx / 3 * np.sum(y[0:-1:2] + 4 * y[1::2] + y[2::2])
    
    return S
    # ====================

Run and check your function.

In [10]:
S = simpson(f, 0, 1, N=50)
print("The result from Simpson's rule is", S)

The result from Simpson's rule is 0.8556243929705796


### Part 3.5

Compute 

$$\int_0^1 e^{-\frac{x^2}{2}} dx$$

with the Gauss-Legendre formula using $n = 1$.

In [11]:
def gauss(
    f,
    n,
    a,
    b
):
    '''
    Parameters
    ----------
    f : function
        Vectorized function of a single variable
    n : integer
        Number of points
    a , b : numbers
        Interval of integration [a,b]

    Returns
    -------
    G : float
        Approximation of the integral of f(x) from a to b using the
        Gaussian–Legendre quadrature rule with N points.
    '''
    # ===== 請實做程式 =====
    [x,w] = p_roots(n)
    G = 0.5*(b-a)*sum(w*f(0.5*(b-a)*x + 0.5*(b+a)))
    # ====================

    return G

Run and check your function.

In [12]:
G = gauss(f, 1, 0, 1)
print("The result from Gauss-Legendre is", G)

The result from Gauss-Legendre is 0.8824969025845955


### Part 3.6

Compare the obtained results of three methods above and write down your observation. You can use either code or markdown to depict.

> Please write down your answer here.

In [18]:
error_1 = abs(G-S)
error_2 = abs(approximation-S)
error_3 = abs(G-approximation)

print('Gaussian:',G, ',Simpson:',S, ',Approximation:',approximation)
error_1, error_2, error_3

Gaussian: 0.8824969025845955 ,Simpson: 0.8556243929705796 ,Approximation: 0.8688435532375445


(0.026872509614015883, 0.013219160266964902, 0.01365334934705098)

The error between Gaussian-Legendre formula and Simpson's rule is the largest.\
The error between the approximation and Simpson's rule is the smallest.\
We can predict that the exact value is between 0.8556243929705796 and 0.8688435532375445.